File: Index code/NBinomRegress.cpp
// Model for whales
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
    // Read data
    DATA_MATRIX(X); // Data matrix
    DATA_MATRIX(Xd); // Dispersion matrix
    DATA_MATRIX(Xhat); // Prediction matrix
    DATA_INTEGER(Nyr); // Number of years
    DATA_VECTOR(P_t); // Probability whale remains in area
    DATA_VECTOR(yobs);
    DATA_INTEGER(Eday); // End day for ingress/egress
    DATA_INTEGER(Iday); // Day of ingress
    
    // Parameters
    PARAMETER_VECTOR(beta); // Expected value parameters
    PARAMETER_VECTOR(betad); // dispersion parameters
    
    // Variables
    vector<Type> eta = X*beta; // Expected value parameters
    vector<Type> etad = Xd*betad; // Dispersion
    
    // Run first stage estimation
    Type jnll = 0;
    for (int i=0; i < yobs.size(); i++){
        jnll -= dnbinom_robust(yobs(i), eta(i), 2. * eta(i) - etad(i), true);
        // jnll -= dnbinom2(yobs(i), eta(i), eta(i) * (1 + etad(i)), true);
    }
    
    // Run second stage calculations
    vector<Type> W_t = exp(Xhat*beta); // Estimated number of whales on day t: W_0 = 0
    vector<Type> dW_t = W_t.size(); dW_t.setZero(); // Difference in whales ingressing on day t: W_0 = 0
    vector<Type> A_xyLong = W_t.size(); A_xyLong.setZero(); // Estimated number of whales on day t: W_0 = 0
    vector<Type> A_xy(Nyr); A_xy.setZero();
    Iday = Iday - 1; // Start of ingress - change for cpp indexing
    Eday = Eday - 1; // stop of ingress/egress - change for cpp indexing
    int Nday = 365; // Number of days
    Type tmp_sum = 0;
    
    
    // -- Should be as follows:
    // -- A_0 = 0
    // -- A_1 = dW_1 + p1 * A_0
    // -- A_2 = dW_2 + p1 * A_1 + p2 * A_0
    // -- A_3 = dW_3 + p1 * A_2 + p2 * A_1 + p3 * A_0
    for(int yr = 0; yr < Nyr; yr++){ // Year loop
        for(int k = 0; k < Nday; k++){ // Day loop - Start at day 100 because W_t = 0 for days 0-99. Indexing starts at 0
            tmp_sum = 0; // Initialize
            
            // No whales before April (day 100, index 99)
            if(k < Iday){
                W_t[Nday*yr + k] = 0;
                dW_t[Nday*yr + k] = 0;
                A_xyLong[Nday*yr + k] = 0;
            }
            
            // Accumulation
            if(k >= Iday & k <= Eday){
                dW_t[Nday*yr + k] = W_t[Nday*yr + k] - W_t[Nday*yr + k-1]; // Calculate dW_t
                
                // Run the summation of all previous days (no way to vectorize)
                for(int ktmp = 0; ktmp < k; ktmp++){
                    tmp_sum += dW_t[Nday*yr + ktmp] + P_t(ktmp) * A_xyLong[Nday*yr + k - ktmp - 1];
                }
                
                A_xyLong[Nday*yr + k] = tmp_sum + dW_t[Nday*yr + k]; // segment of k elements, starting at year
            }
            
            // No whales ingressing after day 320 (index 319)
            if(k > Eday){
                dW_t[Nday*yr + k] = 0;
                A_xyLong[Nday*yr + k] = A_xyLong[Nday*yr + Eday];
            }
            
            // Save value
            if(k == Eday){
                A_xy[yr] = A_xyLong[Nday*yr + Eday];
            }
        }
    }  
    
    // Report predictions
    REPORT( W_t );
    REPORT( dW_t );
    REPORT( A_xyLong );
    REPORT( A_xy );
    ADREPORT( A_xy );
    
    return jnll;
}

File: Index code/NBinomRegress.R
################################################################################
# Load libraries
################################################################################
rm(list=ls())
require(MASS)
library(glmmTMB)
library(TMB)
library(StateSpaceSIR)

################################################################################
# Read in and clean data
################################################################################
balle<- read.csv ( "Data/ballena3.csv" , sep=";" , dec=".")


balle<- balle [c(-11,-12),] #11 y 12 (2004) vuelos con el aerocomander// esta es la seleccion de datos que hice para la los paràmetros para el nùmero de ballenas que dan la vuelta anualmente por PV
balle <- subset (balle, Year<2020) # Subset years

####Variable Respuesta ############
balle$RTA <- (balle$T) # Total of observed whales
balle$Year <- as.factor(balle$Year)


################################################################################
# First stage - regression model
################################################################################
#	Regresion model selected (up to  2019) using Mass package
regresion.balle.nb.jul.cuad <- glm.nb(RTA ~ Year + Juliano, data = balle, link = log)
regresion.balle.glmmTMB.jul.cuad <- glmmTMB(RTA ~ Year + Juliano, data = balle, family = nbinom2(link = "log")) # Doesnt converge
summary (regresion.balle.nb.jul.cuad)
nb.Jul.cuad <- cbind(Estimate = coef(regresion.balle.nb.jul.cuad))
nb.Jul.cuad # Estimates

# Run in TMB to get SD of stage two
setwd("R code")
compile("NBinomRegress.cpp")
dyn.load(dynlib("NBinomRegress"))
setwd("../")

# Set up predictions matrix'
form <- formula(~ Year + Juliano) # Whats the model look like
Xhat <- merge(data.frame(Juliano = 1:365), data.frame(Year = unique(balle$Year)), all = TRUE)
XhatMat <- model.matrix(form, data = Xhat) # Make into matrix

# Set up data, params and estimate
X = model.matrix(form, data = balle) # Make into matrix
data <- list(yobs = balle$RTA, # Response
             Xd=matrix(1, nrow = nrow(balle), ncol = 1), # Dispersion matrix
             X = X, # Design matrix
             Xhat = XhatMat, # Prediction matrix for calculating A_xy
             Nyr = length(unique(balle$Year)), # Number of years
             P_t = dnorm(1:365, 60, 8.66, FALSE), # Probability whale remains in area
             Iday = 100, # Day beginning Ingress
             Eday = 320 # Day ending ingress/egress
             )
parameters <- list(beta=rep(0, ncol(X)), betad=1)
obj <- MakeADFun(data, parameters, DLL="NBinomRegress", hessian = TRUE)
opt <- do.call("optim", obj)


# Compare parameter estimates
Param_est <- data.frame(Name = rownames(nb.Jul.cuad), Mass = round(as.numeric(nb.Jul.cuad), 6), TMB = round(opt$par[1:18], 6))
Param_est$Dif = Param_est$Mass - Param_est$TMB
Param_est # Great! Very close



################################################################################
# Second Stage - accumulated number of whales
################################################################################
# -- Run for years of interest and using nb regression model
A_xy <- data.frame(Year = sort(unique(balle$Year)), B = c(0, nb.Jul.cuad[2:17])) # Years which we want to calculate the accumulated number of whales from years which we have data and the associated regression parameters. NOTE: 1990 is the intercept 

# Get A_xy using MASS and R
A_xy$A_xy <- NA
for(i in 1:nrow(A_xy)){
  A_xy$A_xy[i] <- accum_fun(a = nb.Jul.cuad[1], # Intercept
                            b = A_xy$B[i], # Year parameter
                            c = nb.Jul.cuad[18], # Julian day parameter
                            d = 0, # Julian day^2 parameter
                            mu = 60, # mu from manuscript
                            sigma = 8.66, # sigma from manuscript
                            x = 320,
                            ReportList = FALSE) # Calculate until day 320
  
}
A_xy


# Get A_xy using TMB
report <- obj$report(obj$env$last.par.best)
rep <- sdreport(obj)
Xhat$A_xy <- report$A_xyLong
A_xyTMB <- Xhat[which(Xhat$Juliano == 320),]
A_xyTMB
